Do not add redundant subquery ordering into plan

mertak-synnada opened the pull request:
## Which issue does this PR close?

-


## Rationale for this change

This PR solves an issue with the subquery sortings. If a main query does not support the same sorting with the subquery and the subquery has no limit, then the ordering is redundant (unless used in an Aggregation but it's being handled in a different plan type anyway).

While implementing this behavior we discovered that it's only possible to use `WITH ORDER` syntax with `EXTERNAL` tables. That's why we needed to add `LIMIT xx` conditions to subqueries so that the system can understand they're valid orderings. The sql-parser issue will be handled in a different PR to the sql-parser project and those tests will be changed back to their original forms

Other than that, we discovered there are some redundant LimitExec plans even if the child plan has fetch support. Also, fixed this issue and replaced `GlobalLimitExec` plans with `skip=0` with `LocalLimitExec` plan type.

Also added Limit push down support to the `UnionExec` plan type.

## What changes are included in this PR?

Explained above

## Are these changes tested?

Yes

## Are there any user-facing changes?

No

Commit:
do not add redundant subquery ordering into plan

Commit:
format code

Commit:
add license

Commit:
fix test cases with sort plan removing

mertak-synnada commented on datafusion/sqllogictest/test_files/select.slt:
```diff
@@ -1455,29 +1455,6 @@ physical_plan
 05)--------RepartitionExec: partitioning=RoundRobinBatch(2), input_partitions=1
 06)----------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/core/tests/data/window_2.csv]]}, projection=[a0, a, b, c, d], output_ordering=[a@1 ASC NULLS LAST, b@2 ASC NULLS LAST, c@3 ASC NULLS LAST], has_header=true
 
-# When ordering lost during projection, we shouldn't keep the SortExec.
```
Removed this case since we started to remove the subquery sorting already

Commit:
fix comment

Commit:
keep sorting on ordering mode test cases

Commit:
protect test intentions with order + limit

Commit:
protect test intentions with order + limit

ozankabak commented on datafusion/sql/src/relation/mod.rs:
```diff
@@ -16,6 +16,7 @@
 // under the License.
 
 use crate::planner::{ContextProvider, PlannerContext, SqlToRel};
+use datafusion_common::tree_node::{Transformed, TreeNode};
```
```suggestion
use crate::planner::{ContextProvider, PlannerContext, SqlToRel};

use datafusion_common::tree_node::{Transformed, TreeNode};
```

ozankabak commented on datafusion/sqllogictest/test_files/group_by.slt:
```diff
@@ -2527,20 +2527,24 @@ EXPLAIN SELECT s.country, ARRAY_AGG(s.amount ORDER BY s.amount DESC) AS amounts,
 02)--AggregateExec: mode=Single, gby=[country@0 as country], aggr=[array_agg(s.amount) ORDER BY [s.amount DESC NULLS FIRST], sum(s.amount)], ordering_mode=Sorted
 03)----SortExec: expr=[country@0 ASC NULLS LAST,amount@1 DESC], preserve_partitioning=[false]
-04)------MemoryExec: partitions=1, partition_sizes=[1]
+04)------SortExec: TopK(fetch=10), expr=[country@0 ASC NULLS LAST], preserve_partitioning=[false]
```
One `SortExec` with `TopK(fetch = 10)` and `expr=[country@0 ASC NULLS LAST, amount@1 DESC]` is enough. I think this uncovered a small bug in `EnforceSorting`. Let's fix that too.

ozankabak commented on datafusion/sqllogictest/test_files/group_by.slt:
```diff
@@ -2563,20 +2567,23 @@ EXPLAIN SELECT s.country, s.zip_code, ARRAY_AGG(s.amount ORDER BY s.amount DESC)
 02)--AggregateExec: mode=Single, gby=[country@1 as country, zip_code@0 as zip_code], aggr=[array_agg(s.amount) ORDER BY [s.amount DESC NULLS FIRST], sum(s.amount)], ordering_mode=PartiallySorted([0])
 03)----SortExec: expr=[country@1 ASC NULLS LAST,amount@2 DESC], preserve_partitioning=[false]
-04)------MemoryExec: partitions=1, partition_sizes=[1]
+04)------SortExec: TopK(fetch=10), expr=[country@1 ASC NULLS LAST], preserve_partitioning=[false]
```
Similar to my other comment (possible `EnforceSorting` issue).

ozankabak commented on datafusion/sqllogictest/test_files/group_by.slt:
```diff
@@ -2634,20 +2643,24 @@ EXPLAIN SELECT s.country, ARRAY_AGG(s.amount ORDER BY s.country DESC, s.amount D
 02)--AggregateExec: mode=Single, gby=[country@0 as country], aggr=[array_agg(s.amount) ORDER BY [s.country DESC NULLS FIRST, s.amount DESC NULLS FIRST], sum(s.amount)], ordering_mode=Sorted
 03)----SortExec: expr=[country@0 ASC NULLS LAST,amount@1 DESC], preserve_partitioning=[false]
-04)------MemoryExec: partitions=1, partition_sizes=[1]
+04)------SortExec: TopK(fetch=10), expr=[country@0 ASC NULLS LAST], preserve_partitioning=[false]
```
Similar to my other comment (possible `EnforceSorting` issue).

ozankabak commented on datafusion/sqllogictest/test_files/group_by.slt:
```diff
@@ -2527,20 +2527,24 @@ EXPLAIN SELECT s.country, ARRAY_AGG(s.amount ORDER BY s.amount DESC) AS amounts,
 03)----SubqueryAlias: s
-04)------Sort: sales_global.country ASC NULLS LAST
-05)--------TableScan: sales_global projection=[country, amount]
+04)------Limit: skip=0, fetch=10
+05)--------Sort: sales_global.country ASC NULLS LAST, fetch=10
```
Why do we have an explicit `Limit` if we have a `Sort` with `fetch`?

ozankabak replied:
This is not a bug, resolving.

ozankabak commented on datafusion/sqllogictest/test_files/group_by.slt:
```diff
@@ -2563,20 +2567,23 @@ EXPLAIN SELECT s.country, s.zip_code, ARRAY_AGG(s.amount ORDER BY s.amount DESC)
 03)----SubqueryAlias: s
-04)------Sort: sales_global.country ASC NULLS LAST
-05)--------TableScan: sales_global projection=[zip_code, country, amount]
+04)------Limit: skip=0, fetch=10
+05)--------Sort: sales_global.country ASC NULLS LAST, fetch=10
```
Similar to my other comment (`Limit` and `Sort` with `fetch` at the same time).

ozankabak commented on datafusion/sqllogictest/test_files/group_by.slt:
```diff
@@ -2599,19 +2606,21 @@ EXPLAIN SELECT s.country, ARRAY_AGG(s.amount ORDER BY s.country DESC) AS amounts
 03)----SubqueryAlias: s
-04)------Sort: sales_global.country ASC NULLS LAST
-05)--------TableScan: sales_global projection=[country, amount]
+04)------Limit: skip=0, fetch=10
+05)--------Sort: sales_global.country ASC NULLS LAST, fetch=10
```
Similar to my other comment (`Limit` and `Sort` with `fetch` at the same time).

ozankabak commented on datafusion/sqllogictest/test_files/group_by.slt:
```diff
@@ -2634,20 +2643,24 @@ EXPLAIN SELECT s.country, ARRAY_AGG(s.amount ORDER BY s.country DESC, s.amount D
 03)----SubqueryAlias: s
-04)------Sort: sales_global.country ASC NULLS LAST
-05)--------TableScan: sales_global projection=[country, amount]
+04)------Limit: skip=0, fetch=10
+05)--------Sort: sales_global.country ASC NULLS LAST, fetch=10
```
Similar to my other comment (`Limit` and `Sort` with `fetch` at the same time).

ozankabak commented on datafusion/sqllogictest/test_files/joins.slt:
```diff
@@ -3911,23 +3912,26 @@ EXPLAIN SELECT * FROM (
 04)--SubqueryAlias: rhs
-05)----Sort: right_table_no_nulls.b ASC NULLS LAST
-06)------TableScan: right_table_no_nulls projection=[a, b]
+05)----Limit: skip=0, fetch=10
+06)------Sort: right_table_no_nulls.b ASC NULLS LAST, fetch=10
```
Similar to my other comment (`Limit` and `Sort` with `fetch` at the same time).

ozankabak commented on datafusion/sqllogictest/test_files/joins.slt:
```diff
@@ -4025,18 +4029,20 @@ EXPLAIN SELECT * FROM (
 04)--SubqueryAlias: rhs
-05)----Sort: right_table_no_nulls.b ASC NULLS LAST
-06)------TableScan: right_table_no_nulls projection=[a, b]
+05)----Limit: skip=0, fetch=10
+06)------Sort: right_table_no_nulls.b ASC NULLS LAST, fetch=10
```
Similar to my other comment (`Limit` and `Sort` with `fetch` at the same time).

ozankabak commented on datafusion/sqllogictest/test_files/select.slt:
```diff
@@ -1461,22 +1461,26 @@ query TT
-03)----Sort: aggregate_test_100.c1 ASC NULLS LAST, aggregate_test_100.c2 ASC NULLS LAST
-04)------Projection: aggregate_test_100.c2, aggregate_test_100.c1
-05)--------TableScan: aggregate_test_100 projection=[c1, c2]
+03)----Limit: skip=0, fetch=4
+04)------Sort: aggregate_test_100.c1 ASC NULLS LAST, aggregate_test_100.c2 ASC NULLS LAST, fetch=4
```
Similar to my other comment (`Limit` and `Sort` with `fetch` at the same time).

ozankabak commented on datafusion/sqllogictest/test_files/select.slt:
```diff
@@ -1461,22 +1461,26 @@ query TT
 05)--------RepartitionExec: partitioning=RoundRobinBatch(2), input_partitions=1
-06)----------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/testing/data/csv/aggregate_test_100.csv]]}, projection=[c2], has_header=true
+06)----------ProjectionExec: expr=[c2@0 as c2]
+07)------------SortExec: TopK(fetch=4), expr=[c1@1 ASC NULLS LAST,c2@0 ASC NULLS LAST], preserve_partitioning=[false]
```
This sort could have been a `Limit`, no? Why is it not a `Limit` but a `Sort` with `fetch`? The old plan doesn't have any sorting in the physical plan.

ozankabak replied:
This is not a bug, resolving.

ozankabak commented on datafusion/sqllogictest/test_files/window.slt:
```diff
@@ -1767,27 +1767,25 @@ logical_plan
 02)--AggregateExec: mode=Final, gby=[], aggr=[count(*)]
 03)----CoalescePartitionsExec
 04)------AggregateExec: mode=Partial, gby=[], aggr=[count(*)]
-05)--------RepartitionExec: partitioning=RoundRobinBatch(2), input_partitions=2
```
This seems like an unnecessary repartition that shouldn't have been here in the first place. How did our removal of the logical `Sort` result in this being absent in the new physical plan?

We may have discovered a small bug in `EnforceDistribution` (or some other physical rule) that we should fix. Let's investigate.

ozankabak commented on datafusion/sqllogictest/test_files/subquery_sort.slt:
```diff
@@ -0,0 +1,152 @@
+LOCATION '../../testing/data/csv/aggregate_test_100.csv'
+OPTIONS ('format.has_header' 'true');
+
+# Remove the redundant ordering in the sub query
```
```suggestion
# Remove the redundant ordering in the subquery
```

ozankabak commented on datafusion/sql/src/relation/mod.rs:
```diff
@@ -142,10 +143,36 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
 }
+
+fn optimize_subquery_sort(plan: LogicalPlan) -> Result<Transformed<LogicalPlan>> {
+    // When a subquery is initialized, we look for the sort options since they might be redundant.
+    // It's only important if the subquery result is affected with the order by statement.
+    // Which are the cases of:
+    // 1. DISTINCT ON / ARRAY_AGG ... => It's handled as an Aggregate plan and keep the sorting
+    // 2. RANK / ROW_NUMBER ... => It's handled as a WindowAggr plan sorting is preserved
+    // 3. LIMIT => It's handled as a Sort plan type so that we need to search for it
```
```suggestion
    // When a subquery is initialized, we look for the sort options since they might be unnecessary.
    // It's only important if the subquery result is affected by the ORDER BY statement, which can
    // happen when we have:
    // 1. DISTINCT ON / ARRAY_AGG ... => It's handled as an Aggregate plan and keep the sorting
    // 2. RANK / ROW_NUMBER ... => It's handled as a WindowAggr plan sorting is preserved
    // 3. LIMIT => It's handled as a Sort plan type so that we need to search for it
```

ozankabak reviewed (commented):
Let's address my comments and iterate.

mustafasrepo committed:
Tmp

mustafasrepo committed:
Minor changes

mustafasrepo committed:
Minor changes

Commit:
Merge remote-tracking branch 'refs/remotes/origin/bug_fix/enforce_sorting' into feature/Sort-For-Subqueries

# Conflicts:
#	datafusion/sqllogictest/test_files/window.slt

mustafasrepo committed:
Minor changes

mustafasrepo committed:
Implement top down recursion with delete check

mustafasrepo committed:
Minor changes

Commit:
Merge remote-tracking branch 'refs/remotes/origin/bug_fix/enforce_sorting' into feature/Sort-For-Subqueries

mustafasrepo committed:
Minor changes

Commit:
Merge remote-tracking branch 'refs/remotes/origin/bug_fix/enforce_sorting' into feature/Sort-For-Subqueries

Commit:
initialize fetch() api for execution plan
remove unnecessary limit plans when used with sort + fetch
add test case for Sort and Limit with offset
push down limit even if a child with no fetch appears when the child supports push down

mustafasrepo committed:
Address reviews

mustafasrepo committed:
Update comments

mustafasrepo committed:
Minor changes

mustafasrepo committed:
Make test deterministic

Commit:
add supports limit push down to union exec

Commit:
support limit push down with multi children cases

Commit:
fix typos

Co-authored-by: Mehmet Ozan Kabak <ozankabak@gmail.com>

mustafasrepo committed:
Add fetch info to the statistics

Commit:
optimize tpch test plans

mustafasrepo committed:
Enforce distribution use inexact count estimate also.

mustafasrepo committed:
Minor changes

mustafasrepo committed:
Minor changes

Commit:
Merge remote-tracking branch 'refs/remotes/origin/bug_fix/enforce_sorting' into feature/Sort-For-Subqueries

# Conflicts:
#	datafusion/sqllogictest/test_files/aggregates_topk.slt

Commit:
Merge branch 'refs/heads/bug_fix/enforce_sorting' into feature/Sort-For-Subqueries

Commit:
Merge remote-tracking branch 'refs/remotes/origin/apache_main' into feature/Sort-For-Subqueries

# Conflicts:
#	datafusion/core/src/physical_optimizer/enforce_distribution.rs
#	datafusion/core/src/physical_optimizer/enforce_sorting.rs
#	datafusion/core/src/physical_optimizer/limit_pushdown.rs
#	datafusion/core/src/physical_optimizer/sort_pushdown.rs
#	datafusion/physical-plan/src/coalesce_batches.rs
#	datafusion/physical-plan/src/limit.rs
#	datafusion/physical-plan/src/sorts/sort.rs
#	datafusion/sqllogictest/test_files/order.slt

Commit:
merge with apache main
add pushes_global_limit_into_multiple_fetch_plans test case
change limit_pushdown.rs as manual top down operator and simplify algorithm by supporting most parent node remove and other pushdown cases

Commit:
format code

Commit:
fix doc paths

Commit:
fix doc paths

Commit:
Merge branch 'refs/heads/apache_main' into feature/Sort-For-Subqueries

Commit:
remove redundant code block

ozankabak commented on datafusion/physical-plan/src/union.rs:
```diff
@@ -260,6 +260,13 @@ impl ExecutionPlan for UnionExec {
     }
+
+    fn supports_limit_pushdown(&self) -> bool {
+        // UnionExec is supporting limit_pushdown because it's only applicable with
+        // SortPreservingMergeExec plan. Since SortExec does not support limit_pushdowm
+        // it's not possible to lose sorting in UnionExec's children.
```
I don't understand what this is trying to say

mertak-synnada replied:
We discussed adding the supports_limit_pushdown method to UnionExec or not. We agreed to do so because LimitPushDown is only applicable to SPM in UnionExec's possible children. That's why I put this comment actually, maybe we can put a better comment I don't know

ozankabak commented on datafusion/sql/src/relation/mod.rs:
```diff
@@ -142,10 +144,39 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
+    // happen when we have:
+    // 1. DISTINCT ON / ARRAY_AGG ... => It's handled as an Aggregate plan and keep the sorting
+    // 2. RANK / ROW_NUMBER ... => It's handled as a WindowAggr plan sorting is preserved
+    // 3. LIMIT => It's handled as a Sort plan type so that we need to search for it
```
```suggestion
    // When initializing subqueries, we examine sort options since they might be unnecessary.
    // They are only important if the subquery result is affected by the ORDER BY statement,
    // which can happen when we have:
    // 1. DISTINCT ON / ARRAY_AGG ... => Handled by an `Aggregate` and its requirements.
    // 2. RANK / ROW_NUMBER ... => Handled by a `WindowAggr` and its requirements.
    // 3. LIMIT => Handled by a `Sort`, so we need to search for it.
```

Commit:
if partition count is 1 put GlobalLimitExec

Commit:
fix test cases

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -111,34 +123,171 @@ impl From<LimitExec> for Arc<dyn ExecutionPlan> {
     if let Some(global_limit) = plan.as_any().downcast_ref::<GlobalLimitExec>() {
         Some(LimitExec::Global(GlobalLimitExec::new(
-            Arc::clone(global_limit.input()),
+            Arc::<dyn ExecutionPlan>::clone(global_limit.input()),
```
```suggestion
            Arc::clone(global_limit.input()),
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -147,107 +296,48 @@ fn extract_limit(plan: &Arc<dyn ExecutionPlan>) -> Option<LimitExec> {
             .map(|local_limit| {
                 LimitExec::Local(LocalLimitExec::new(
-                    Arc::clone(local_limit.input()),
+                    Arc::<dyn ExecutionPlan>::clone(local_limit.input()),
```
```suggestion
                    Arc::clone(local_limit.input()),
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -147,107 +296,48 @@ fn extract_limit(plan: &Arc<dyn ExecutionPlan>) -> Option<LimitExec> {
+    let new_plan: Arc<dyn ExecutionPlan> =
+        Arc::new(LocalLimitExec::new(pushdown_plan, fetch));
+    pushdown_plan = new_plan;
+    pushdown_plan
```
```suggestion
    Arc::new(LocalLimitExec::new(pushdown_plan, fetch)) as _
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -147,107 +296,48 @@ fn extract_limit(plan: &Arc<dyn ExecutionPlan>) -> Option<LimitExec> {
+    let new_plan: Arc<dyn ExecutionPlan> =
+        Arc::new(GlobalLimitExec::new(pushdown_plan, skip, fetch));
+    pushdown_plan = new_plan;
+    pushdown_plan
```
```suggestion
    Arc::new(GlobalLimitExec::new(pushdown_plan, skip, fetch)) as _
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -111,34 +123,171 @@ impl From<LimitExec> for Arc<dyn ExecutionPlan> {
-            add_fetch_to_child(&limit_exec, Arc::clone(child))
+/// pushdown_limit_helper function is the main helper function of the `LimitPushDown` rule
+/// The helper takes the `ExecutionPlan` and a global state which is an instance of `GlobalRequirements`
+/// and modifies these parameters while checking if the limits can be pushed down or not.
```
```suggestion
/// This function is the main helper function of the `LimitPushDown` rule.
/// The helper takes an `ExecutionPlan` and a global (algorithm) state which is an instance of
/// `GlobalRequirements` and modifies these parameters while checking if the limits can be
/// pushed down or not.
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -50,7 +65,13 @@ impl PhysicalOptimizerRule for LimitPushdown {
+            satisfied: false,
+        };
+        let updated_plan = pushdown_limits(plan, global_state)?;
+        Ok(updated_plan)
```
```suggestion
        pushdown_limits(plan, global_state)
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -111,34 +123,171 @@ impl From<LimitExec> for Arc<dyn ExecutionPlan> {
+    mut global_state: GlobalRequirements,
+) -> Result<(Transformed<Arc<dyn ExecutionPlan>>, GlobalRequirements)> {
+    if let Some(limit_exec) = extract_limit(&pushdown_plan) {
+        // If we have fetch/skip info in global state already we need to decide which one to continue with
```
```suggestion
        // If we have fetch/skip info in the global state already, we need to decide
        // which one to continue with:
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -111,34 +123,171 @@ impl From<LimitExec> for Arc<dyn ExecutionPlan> {
+        // We will decide later if we should add it again or not
+        return Ok((
+            Transformed {
+                data: Arc::<dyn ExecutionPlan>::clone(limit_exec.input()),
```
```suggestion
        // Now the global state has the most recent information, we can remove the `LimitExec`
        // plan. We will decide later if we should add it again or not.
        return Ok((
            Transformed {
                data: Arc::clone(limit_exec.input()),
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -111,34 +123,171 @@ impl From<LimitExec> for Arc<dyn ExecutionPlan> {
+
+    // The plan is not LimitExec
+    if pushdown_plan.fetch().is_some() {
+        // Plan has its own fetch info, decide which one to use in global_state
```
```suggestion
    // If we have a non-limit operator with fetch capability, update global state as necessary:
    if pushdown_plan.fetch().is_some() {
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -147,107 +296,48 @@ fn extract_limit(plan: &Arc<dyn ExecutionPlan>) -> Option<LimitExec> {
+    if skip > 0 || pushdown_plan.output_partitioning().partition_count() == 1 {
+        return add_global_limit(pushdown_plan, skip, Some(fetch));
+    }
+    add_local_limit(pushdown_plan, fetch + skip)
```
```suggestion
    if skip > 0 || pushdown_plan.output_partitioning().partition_count() == 1 {
        add_global_limit(pushdown_plan, skip, Some(fetch))
    } else {
        add_local_limit(pushdown_plan, fetch + skip)
    }
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -111,34 +123,171 @@ impl From<LimitExec> for Arc<dyn ExecutionPlan> {
+        return Ok((Transformed::no(pushdown_plan), global_state));
+    }
+
+    if pushdown_plan.supports_limit_pushdown() {
+        if !combines_input_partitions(&pushdown_plan) {
+            // We have the info in global state and the plan is pushing down, continue
+            return Ok((Transformed::no(pushdown_plan), global_state));
+        }
```
```suggestion
    if pushdown_plan.supports_limit_pushdown() && !combines_input_partitions(&pushdown_plan) {
        // We have information in the global state and the plan pushes down, continue:
        return Ok((Transformed::no(pushdown_plan), global_state));
    }
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -111,34 +123,171 @@ impl From<LimitExec> for Arc<dyn ExecutionPlan> {
+        );
+    }
+
+    if global_state.fetch.is_none() {
+        // There's no valid Fetch info
+        if global_state.skip > 0 {
+            // There might be a case with only offset, if so add a global limit
+            if !global_state.satisfied {
+                global_state.satisfied = true;
+                return Ok((
+                    Transformed::yes(add_global_limit(
+                        pushdown_plan,
+                        global_state.skip,
+                        global_state.fetch,
+                    )),
+                    global_state,
+                ));
+            }
+        }
+        // There's no info on fetch, nothing to do
+        return Ok((Transformed::no(pushdown_plan), global_state));
+    }
```
```suggestion
    let global_fetch = Some(global_state.fetch) else {
        // There's no valid fetch information, exit early:
        return if global_state.skip > 0 && !global_state.satisfied {
            // There might be a case with only offset, if so add a global limit:
            global_state.satisfied = true;
            Ok((
                Transformed::yes(add_global_limit(
                    pushdown_plan,
                    global_state.skip,
                    None,
                )),
                global_state,
            ));
        } else {
            // There's no info on fetch, nothing to do:
            Ok((Transformed::no(pushdown_plan), global_state))
        }
    };
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -111,34 +123,171 @@ impl From<LimitExec> for Arc<dyn ExecutionPlan> {
+
+        // This plan is combining input partitions we need to add the fetch info to plan if possible
+        // if not, we must add a LimitExec with the global_state info
+        let skip_and_fetch = Some(global_state.fetch.unwrap() + global_state.skip);
```
```suggestion
        let skip_and_fetch = Some(global_fetch + global_state.skip);
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -111,34 +123,171 @@ impl From<LimitExec> for Arc<dyn ExecutionPlan> {
+            return Ok((Transformed::yes(pushdown_plan), global_state));
+        }
+
+        if global_state.satisfied {
+            // If the plan is already satisfied do not add a limit info
+            return Ok((Transformed::no(pushdown_plan), global_state));
+        }
+
+        // Add LimitExec
+        global_state.satisfied = true;
+        return Ok((
+            Transformed::yes(add_limit(
+                pushdown_plan,
+                global_state.skip,
+                global_state.fetch.unwrap(),
+            )),
+            global_state,
+        ));
+    }
```
```suggestion
        return if global_state.satisfied {
            // If the plan is already satisfied, do not add a limit:
            Ok((Transformed::no(pushdown_plan), global_state));
        } else {
            global_state.satisfied = true;
            Ok((
                Transformed::yes(add_limit(
                    pushdown_plan,
                    global_state.skip,
                    global_fetch,
                )),
                global_state,
            ));
        }
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -111,34 +123,171 @@ impl From<LimitExec> for Arc<dyn ExecutionPlan> {
+        ));
+    }
+
+    // The plan does not support push down and not a LimitExec
+    // We will need to add a LimitExec or Fetch info if the plan is not satisfied
+    // If the plan is already satisfied we will try to add the Fetch info and return the plan
+
+    let fetch = global_state.fetch;
+    let skip = global_state.skip;
+    let skip_and_fetch = Some(fetch.unwrap() + skip);
```
```suggestion
    // The plan does not support push down and it is not a limit.
    // We will need to add a limit or a fetch. If the plan is already satisfied,
    // we will try to add the fetch info and return the plan.

    let skip_and_fetch = Some(global_fetch + global_state.skip);
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -111,34 +123,171 @@ impl From<LimitExec> for Arc<dyn ExecutionPlan> {
+    global_state.fetch = None;
+    global_state.skip = 0;
+    if global_state.satisfied {
+        if let Some(plan_with_fetch) = pushdown_plan.with_fetch(skip_and_fetch) {
+            pushdown_plan = plan_with_fetch;
+            return Ok((Transformed::yes(pushdown_plan), global_state));
+        }
+        return Ok((Transformed::no(pushdown_plan), global_state));
```
```suggestion
        return if let Some(plan_with_fetch) = pushdown_plan.with_fetch(skip_and_fetch) {
            Ok((Transformed::yes(plan_with_fetch), global_state));
        } else {
            Ok((Transformed::no(pushdown_plan), global_state));
        }
```

ozankabak commented on datafusion/physical-optimizer/src/limit_pushdown.rs:
```diff
@@ -111,34 +123,171 @@ impl From<LimitExec> for Arc<dyn ExecutionPlan> {
+
+    // Add Fetch / LimitExec
+    global_state.satisfied = true;
+    if let Some(plan_with_fetch) = pushdown_plan.with_fetch(skip_and_fetch) {
+        pushdown_plan = plan_with_fetch;
+        if skip > 0 {
+            return Ok((
+                Transformed::yes(add_global_limit(pushdown_plan, skip, fetch)),
+                global_state,
+            ));
+        }
+        return Ok((Transformed::yes(pushdown_plan), global_state));
+    }
+
+    Ok((
+        Transformed::yes(add_limit(pushdown_plan, skip, fetch.unwrap())),
+        global_state,
+    ))
```
```suggestion
    if let Some(plan_with_fetch) = pushdown_plan.with_fetch(skip_and_fetch) {
        return if skip > 0 {
            Ok((
                Transformed::yes(add_global_limit(plan_with_fetch, skip, fetch)),
                global_state,
            ));
        } else {
            Ok((Transformed::yes(pushdown_plan), global_state));
        };
    }

    Ok((
        Transformed::yes(add_limit(pushdown_plan, skip, global_fetch)),
        global_state,
    ))
```

ozankabak commented on datafusion/physical-plan/src/coalesce_batches.rs:
```diff
@@ -213,7 +213,13 @@ impl ExecutionPlan for CoalesceBatchesExec {
 
     fn statistics(&self) -> Result<Statistics> {
-        Statistics::with_fetch(self.input.statistics()?, self.schema(), self.fetch, 0, 1)
+        Statistics::with_fetch(
+            self.input.statistics()?,
+            self.schema(),
+            self.fetch,
+            0,
+            self.properties().output_partitioning().partition_count(),
+        )
```
```suggestion
        Statistics::with_fetch(self.input.statistics()?, self.schema(), self.fetch, 0, 1)
```

ozankabak commented on datafusion/physical-plan/src/limit.rs:
```diff
@@ -325,7 +325,7 @@ impl ExecutionPlan for LocalLimitExec {
             Some(self.fetch),
             0,
-            1,
+            self.properties().output_partitioning().partition_count(),
```
```suggestion
            1,
```

ozankabak reviewed (commented):
Close to finalizing

ozankabak committed:
Apply suggestions from code review

Commit:
fix syntax errors

ozankabak committed:
Simplify branches

Commit:
Merge branch 'refs/heads/apache_main' into feature/Sort-For-Subqueries

# Conflicts:
#	datafusion/sqllogictest/test_files/window.slt

Commit:
remove redundant limit plans from merge

ozankabak commented:
I am going to merge this today unless there are any concerns raised. We will then proceed with `WITH ORDER` support for memory tables and remove the temporary `LIMIT` clauses introduced

ozankabak commented:
Merged upstream.
